aqueduct_client/aqueduct_client.py

```diff
@@ -0,0 +1,243 @@
+
+        Parameters
+        ----------
+        pipeline_id : str
+            The id of the pipeline whose results should be loaded.
+
+        Returns
+        -------
+        pd.DataFrame
+            A dataframe holding the result, indexed by date and the
+            asset identifier format (symbol, sid, or fsym_region_id)
+            that this pipeline used.
+        """
+        pipeline_status = self.get_pipeline(pipeline_id)
+        if pipeline_status["status"] == "RUNNING":
+            raise ValueError("Pipeline {id} is still running!".format(
+                id=pipeline_id
+            ))
+        elif pipeline_status["status"] == "FAILED":
+            raise ValueError(
+                "Pipeline {id} ended in error, use `get_error` "
+                "to get its error message."
+            ).format(id=pipeline_id)
+
+        asset_identifier_format = pipeline_status["asset_identifier_format"]
+
+        # now that we know the pipeline isn't still running, get its results
+        response = self._get('/{id}/results_url'.format(
+            base=self._base_url,
+            id=pipeline_id
+        ), None)
+
+        url = response.json()['url']
+
+        # get the data from the url
+        results_url_resp = requests.get(url)
+
+        if results_url_resp.status_code != 200:
+            raise ValueError("Could not download results from given url.")
+
+        result_df = pd.read_csv(
+            pd.compat.StringIO(results_url_resp.text),
+            index_col=['date', asset_identifier_format],
+        )
+
+        return result_df
+
+    def get_error(self, id):
+        """
+        Gets the error that caused this pipeline to fail to complete
+        successfully.
+
+        Parameters
+        ----------
+        id : str
+            The id of the pipeline whose errors should be loaded.
+
+        Returns
+        -------
+        dict
+            A dictionary that can contain `date`, `name`, `message`,
+            `lineno`, `method` keys.
+        """
+        pipeline_status = self.get_pipeline(id)
+        if pipeline_status["status"] != "FAILED":
+            raise ValueError(
+                "Pipeline {id} did not end in error!"
+            ).format(id=id)
+
+        # get the error
+        response = self._get(
+            '/{id}/exception'.format(id=id),
+            None
+        )
+
+        response.raise_for_status()
+
+        return response.json()
+
+    def _get(self, path, body, *args, **kwargs):
+        return requests.get(
+            self._base_url + path,
+            *args,
+            headers={'Quantopian-API-Key': self._api_key},
+            params=body,
+            **kwargs
+        )
+
+    def _post(self, path, body, *args, **kwargs):
+        return requests.post(
+            self._base_url + path,
+            *args,
+            headers={'Quantopian-API-Key': self._api_key},
+            json=body,
+            **kwargs
+        )
```

Comment:
It seems like somewhere we probably want a method that submits a Pipeline, polls until it's ready, and then returns the result once it's ready? There are advanced use cases that will want these lower-level primitives, but I'd expect the 90+% use-case to be just calling `run_pipeline(string, start_date, end_date)` with the above semantics. I'd even go so far as to say that we probably want a top-level API method called `run_pipeline` whose implementation would look something like:

```python
def run_pipeline(pipeline, start_date, end_date, api_key=None):
    client = create_aqueduct_client(api_key=api_key)
    return client.run_pipeline(
        pipeline=pipeline,
        start_date=start_date,
        end_date=end_date,
    )
```

We could also use the above to easily implement an IPython/Jupyter magic, which would let users just do:

```
%% aqueduct -s 2014 -e 2015 --api-key FOOBAR (optional)
from quantopian.pipeline import Pipeline, ...

def make_pipeline():
    return Pipeline(...)
```

Reply:
Agreed. I'd like to do this in a subsequent version (soon) - ok with you?

The "poll until ready" method would benefit from some sort of way to get progress, because we can still be waiting a couple of minutes for a pipeline execution to finish - so I like letting the user be able to do other stuff while waiting.